### PR TITLE
Added constructors to immediately open an X509Store

### LINKED ISSUE #6550: System.Security.Cryptography.X509Certificates.X509Store hides errors when cert collection is obtained from a closed store
`System.Security.Cryptography.X509Certificates.X509Store` has

```
     public X509Certificate2Collection Certificates { get {
```

which silently returns an empty collection if the store is not opened. The code to traverse a store looks something like this:

```
 using( var store = new X509Store() ) {
       store.Open( OpenFlags.ReadOnly );
       var allCerts = store.Certificates;
       //traverse allCerts
 }
```

if `Open()` call is omitted for any reason - "no problem", the code works, just looks at empty collection. This is error hiding. `Add()` and `Remove()` don't allow the store to be not opened in advanced. There's no way to check the store is opened. Something like this:

```
  void traverseCerts( X509Store store ) {
       var allCerts = store.Certificates;
       //traverse allCerts
  }
```

cannot even check the store is opened - it gets a collection anyway, maybe empty and it's empty maybe because the store is not opened or because the store is indeed empty.

The same behavior is on desktop. This is extremely inconvenient. A `IsOpened` property is surely needed. Also `Certificates` getter likely needs to throw an exception if the store is not opened in advance.

### PR DESCRIPTION
**Pull request by jmezach**
In this PR new constructor overloads have been added to the X509Store class which immediately open the store, instead of having to call Open on the constructed instance. Additionally an IsOpen property has been added to the X509Store that returns true or false depending on whether the store is open or not.

Fixes #6550

### CODE REVIEW

**Comment by jmezach** [comment]
Interesting. It looks like Open() with a non-existing store throws a CryptographicException on all platforms except for Windows, which causes my tests to fail on Windows. On Windows I get the following exception:

Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException

Is that by design?

**Comment by karelz** [comment]
@bartonjs can you please assess risk for ZBB on 5/10. If it is risky, let's delay merging the PR until after 5/10 (let me know and I will mark the PR appropriately)

Here's details why: https://github.com/dotnet/corefx/issues/17619#issuecomment-296872132

**Comment by krwq** [comment]
WindowsCryptographicException inherits from CryptographicException: https://github.com/dotnet/corefx/blob/c55435befeb372707558a5c441b3c64bf243948f/src/Common/src/Internal/Cryptography/Windows/CryptoThrowHelper.cs#L19 - I think this is correct

**Comment by jmezach** [comment]
Ah, right, I should have used ThrowsAny of course. Will fix that.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Store.cs [hunk 1]
```diff
@@ -70,6 +70,12 @@ public X509Store(StoreName storeName, StoreLocation storeLocation)
             Location = storeLocation;
         }
 
+        public X509Store(StoreName storeName, StoreLocation storeLocation, OpenFlags flags)
+            : this(storeName, storeLocation)
+        {
+            this.Open(flags);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Store.cs]: Style: We don't use `this.` unless the code won't compile without it.

File: src/System.Security.Cryptography.X509Certificates/tests/X509StoreTests.cs [hunk 2]
```diff
@@ -88,6 +97,32 @@ public static void Constructor_StoreHandle_Unix()
...
+                Assert.True(store.IsOpen);
+            }
+        }
+
+        [Fact]
+        public static void Constructor_OpenFlags_StoreName()
+        {
+            using (X509Store store = new X509Store("My", StoreLocation.CurrentUser, OpenFlags.ReadOnly))
+            {
+                Assert.True(store.IsOpen);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/X509StoreTests.cs]: Please add some more tests:

* Using the current ctors and calling Open directly, is IsOpen correct.
* Call Close or Dispose on an object still in scope, does IsOpen go to false?
* Call Open again after closing it, is IsOpen true again?
* Use the new ctor, then call Open again, does IsOpen stay open?

**Comment by bartonjs** [comment]
@karelz With the additional tests I don't think that I'd call this risky.  If we're generally at API freeze then we can hold; but this barely registers on my worry-o-meter.

**Comment by bartonjs** [comment]
Late API Triage:

There are two parts to this feature:
* `IsOpen`: Allows for inspection of the object state.
  * For better or worse the store.Certificates property returns an empty collection when the store is not in an open state.  `IsOpen` allows for disambiguation on the empty collection (or enables API to throw if it's not open, et cetera).
* The new ctors: Convenience/clarification API
  * Many people, when first encountering X509Store, fail to identify that they need to call Open.  The new ctors accept the OpenFlags directly and allow for more compact code, as well as give a general indication that the flags should be provided.

Since IsOpen is just exposing state that's already tracked, and the new ctors are just calling already public API, the change is low risk.

**Comment by karelz** [comment]
@Petermarcu @danmosemsft I would like to consider this for 2.0 given the reasonably high value and low risk (once code review feedback is incorporated).
Thoughts?

**Comment by karelz** [comment]
cc @stephentoub

**Comment by danmosemsft** [comment]
If this change ~cannot change existing behavior, which appears to be the case, I have no objection to 2.0.

**Comment by jmezach** [comment]
@bartonjs I've added the requested tests and removed the unnecessary this in the constructor.

**Comment by jmezach** [comment]
@karelz Is this ready to merge, or are we still waiting for something?

**Comment by karelz** [comment]
Sounds ready for merge to me ... @bartonjs?

**Comment by bartonjs** [comment]
@dotnet-bot Test Innerloop Ubuntu16.04 arm Debug Cross Build please (github timeout)

**Comment by bartonjs** [comment]
@karelz I thought we were waiting on @Petermarcu to approve the late API change.

**Comment by karelz** [comment]
Ah, missed that. Can you please send email to .NET Core shiproom? That will get traction for sure ;-)

**Comment by danmosemsft** [comment]
@Petermarcu approved all these API provided they are added before 5/10 fork

https://github.com/dotnet/corefx/issues/13800  
https://github.com/dotnet/corefx/issues/17051 
https://github.com/dotnet/corefx/issues/16836
https://github.com/dotnet/corefx/issues/4865 
https://github.com/dotnet/corefx/pull/19145

**Pull request merged**